test/extended/operators/workload_guarantees.go

```diff
@@ -0,0 +1,269 @@
+				if tolerates.Key == "node-role.kubernetes.io/master" && tolerates.Operator == corev1.TolerationOpExists {
+					toleratesMaster = true
+				}
+			}
+			if toleratesMaster {
+				continue
+			}
+
+			key := fmt.Sprintf("%s/%s/%s", workload.GetObjectKind().GroupVersionKind().Kind, meta.Namespace, meta.Name)
+
+			if inHAMode {
+				if replicas == 1 {
+					// TODO: We need to check if this is a misconfiguration.
+					e2e.Logf("The %s component has a single replica when OpenShift is configured to be HA", key)
+					continue
+				}
+				workloadStatus, isInvalidWorkload := isInvalidWorkloadInHAMode(knownBugs, key, workload, template)
+				if isInvalidWorkload && workloadStatus != "" {
+					invalidWorkloads = append(invalidWorkloads, workloadStatus)
+				} else if !isInvalidWorkload && workloadStatus != "" {
+					knownBroken = append(knownBroken, workloadStatus)
+				}
+			} else if !inHAMode {
+				// For SNO cases
+				nonHAPrefix := "non-HA/"
+				key = nonHAPrefix + key
+
+				if replicas != 1 {
+					rule := fmt.Sprintf("%s/%s", key, "use_single_replica_non_ha")
+					violation := fmt.Sprintf("%s must set replicas to 1 in non-HA mode", rule)
+					if bug, ok := knownBugs[key]; ok {
+						knownBroken = append(knownBroken, fmt.Sprintf("%s (bug %s)", violation, bug))
+					} else {
+						invalidWorkloads = append(invalidWorkloads, violation)
+					}
+				} else if !isSoftAntiAffinity(template.Spec.Affinity, template.Spec.TopologySpreadConstraints) && replicas >= 2 {
+					rule := fmt.Sprintf("%s/%s", key, "use_soft_anti_affinity_non_ha")
+					violation := fmt.Sprintf("%s must set pod anti-affinity preferred during scheduling, with topologyKey 'kubernetes.io/hostname'", rule)
+					if bug, ok := knownBugs[key]; ok {
+						knownBroken = append(knownBroken, fmt.Sprintf("%s (bug %s)", violation, bug))
+					} else {
+						invalidWorkloads = append(invalidWorkloads, violation)
+					}
+				}
+
+			}
+		}
+
+		e2e.Logf("Workload configuration in payload:")
+
+		// All known bugs are listed as flakes so we can see them as dashboards
+		if len(knownBroken) > 0 {
+			sort.Strings(knownBroken)
+			result.Flakef("Workloads with outstanding bugs in payload:\n%s", strings.Join(knownBroken, "\n"))
+		}
+
+		// Users are not allowed to add new violations
+		if len(invalidWorkloads) > 0 {
+			e2e.Failf("Workloads found that do not meet platform requirements for HA strategy:\n  %s", strings.Join(invalidWorkloads, "\n  "))
+		}
+	})
+})
+
+func isHardAntiAffinity(affinity *corev1.Affinity, spread []corev1.TopologySpreadConstraint) bool {
+	if affinity != nil && affinity.PodAntiAffinity != nil {
+		for _, term := range affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+			if term.TopologyKey == "kubernetes.io/hostname" {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func isSoftAntiAffinity(affinity *corev1.Affinity, spread []corev1.TopologySpreadConstraint) bool {
+	if affinity != nil && affinity.PodAntiAffinity != nil {
```

Comment:
Do we want to fail in non-HA mode when workloads don't define any affinity? i.e. do we care whether it's `Preferred` or just no anti-affinity defined at all?

Reply:
Valid point. I'm modified the test suite to include that scenario.